File: Day_2_1306. Jump Game III (Raj)/Solution.cpp
class Solution
{
public:
  bool canReach(vector<int> &arr, int start)
  {
    // array to keep track of visited posititions
    vector<bool> visited(arr.size(), false);
    // mark starting position as visited
    visited[start] = true;
    // queue for BFS approach
    queue<int> q;
    q.push(start);
    while (q.size() > 0)
    {
      int curr = q.front();
      q.pop();
      if (arr[curr] == 0)
        return true;
      int front = curr + arr[curr];
      int back = curr - arr[curr];
      if (front < arr.size() && !visited[front])
      {
        q.push(front);
        visited[front] = true;
      }
      if (back >= 0 && !visited[back])
      {
        q.push(back);
        visited[back] = true;
      }
    }
    return false;
  }
};

File: Day_2_1306. Jump Game III (Raj)/Solution.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        # array to keep track of visited posititions
        visited = [False]*len(arr)
        # mark starting position as visited
        visited[start] = True
        # queue for BFS approach
        q = deque()
        q.append(start)
        while len(q) > 0:
            curr = q.popleft()
            if arr[curr] == 0:
                return True
            front = curr + arr[curr]
            back = curr - arr[curr]
            if front < len(arr) and not visited[front]:
                q.append(front)
                visited[front] = True
            if back >= 0 and not visited[back]:
                q.append(back)
                visited[back] = True
        return False
